### ISSUE TITLE
bug: Kanshi does not restart with Sway

### ISSUE THREAD

**Issue by houstdav000** [nix-community/home-manager#2797]
### Is there an existing issue for this?

- [X] I have searched the existing issues

### Issue description

Sway can restart without the systemd `kanshi.service` unit restarting, resulting in configuration not being applied correctly after Sway reload, as Kanshi doesn't get notified on reload. In fact, the [Kanshi nixpkgs default.nix](https://github.com/NixOS/nixpkgs/blob/master/pkgs/tools/wayland/kanshi/default.nix) doesn't even include the optional libvarlink dependency, so its IPC hooks won't work anyways. Only way might be to add a conditional append to the Sway (or other Wayland WM) config's OnChange to restart the service with systemctl, since Sway isn't run as a Systemd service anyways.

I've currently fixed it by adding an activation entry as a workaround, but that's hardly a good fix imo.

### Maintainer CC

@nurelin 

### System information

```markdown
- system: `"x86_64-linux"`
 - host os: `Linux 5.16.14, NixOS, 22.05 (Quokka)`
 - multi-user?: `yes`
 - sandbox: `yes`
 - version: `nix-env (Nix) 2.7.0`
 - channels(root): `"nixos"`
 - channels(david): `""`
 - nixpkgs: `/nix/var/nix/profiles/per-user/root/channels/nixos`
```

**Comment by nurelin** [comment]
Yeah, that is not ideal. I am not sure what should be done.

Concerning `kanshictl`, libvarlink is not part of the derivation because the ipc protocol has been added in an unreleased version of kanshi. I was able to build it using:

```
  services.kanshi = {
    enable = true;
    package = (pkgs.kanshi.overrideAttrs (old: {
      buildInputs = old.buildInputs ++ [ pkgs.libvarlink ];
      src = pkgs.fetchFromGitHub {
        owner = "emersion";
        repo = "kanshi";
        rev = "0faad7544323dfade5def5f9e7f132584cc1cf56";
        sha256 = "sha256-AELUyBa3ZCTRd725lRvIbBxwRFiSYiN0SqtasSGXp1I=";
      };
    }));
};
```

For the problem of reloading kanshi when sway reload/restart, I am not sure what would be the best solution.
It will depend on how sway itself is launched. If sway is launched through a systemd's unit then some systemd magic could be worked out but if it is not, then maybe adding:

```
exec systemctl restart kanshi.service
# or
exec kanshictl reload
```

would be the easiest solution. But that would not be very satisfying either.

**Comment by nurelin** [comment]
Yeah, I was hoping to find a more general solution. Ideally, sway itself should signal that kanshi needs to reapply its configuration.
That is why I was proposing to add an exec directly in sway's configuration.

**Comment by houstdav000** [comment]
> Something like:
> 
> ```nix
> # services/kanshi.nix
>   config = mkIf cfg.enable {
>     wayland.windowManager.sway.config.startup = [ 
>       { command = "${pkgs.systemd}/bin/systemctl --user reload-or-restart kanshi"; }
>    ];
>   }
> ```

This is probably the best stopgap measure imo, but since this display output service isn't particularly tied to Sway, just Wayland, it might be better to have some kind of attrset or list of attrsets at the `wayland.windowManager` define a list of general restart/start hooks that `wayland.windoManager.sway.config.startup` could then inherit/join into its own, and include this kanshi reload/restart in.

**Comment by g0hl1n** [comment]
> Something like:
> 
> ```nix
> # services/kanshi.nix
>   config = mkIf cfg.enable {
>     wayland.windowManager.sway.config.startup = [ 
>       { command = "${pkgs.systemd}/bin/systemctl --user reload-or-restart kanshi"; }
>    ];
>   }
> ```

Unfortunately this doesn't work for me on NixOS 23.05 with home-manager 23.05 when running `nixos-rebuild switch` :disappointed:

But running `systemctl --user reload-or-restart kanshi` in a shell afterwards works... Any ideas?

**Comment by MadnessASAP** [comment]
> > Something like:
> > ```nix
> > # services/kanshi.nix
> >   config = mkIf cfg.enable {
> >     wayland.windowManager.sway.config.startup = [ 
> >       { command = "${pkgs.systemd}/bin/systemctl --user reload-or-restart kanshi"; }
> >    ];
> >   }
> > ```
> 
> Unfortunately this doesn't work for me on NixOS 23.05 with home-manager 23.05 when running `nixos-rebuild switch` 😞
> 
> But running `systemctl --user reload-or-restart kanshi` in a shell afterwards works... Any ideas?

add `always = true;` to the startup submodule. i.e.

```nix
{ command = "${pkgs.systemd}/bin/systemctl --user reload-or-restart kanshi.service"; always = true; }
```

This way sway executes the command every time it restarts, not just the first time.

**Comment by Vash1** [comment]
for anyone in the future having the same problem:
```
# ~/.config/sway/config
exec kanshi
exec_always kanshictl reload
```
Im not running kanshi as a systemd service, and if you do, just the kanshictl reload line should work fine.
